src/Accelerators/NNPA/Conversion/ONNXToZHigh/DevicePlacement.cpp

```diff
@@ -93,26 +181,101 @@ void DevicePlacementPass::runOnOperation() {
+    }
+  }
+
+  // Create a JSON configuration file if required.
+  if (!saveConfigFile.empty())
+    saveConfigToJSONFile();
+}
+
+void DevicePlacementPass::loadConfigFromJSONFile() {
+  auto Buf = ExitOnErr(errorOrToExpected(
+      llvm::MemoryBuffer::getFile(loadConfigFile, /*bool IsText=*/true,
+          /*RequiresNullTerminator=*/false)));
+  auto jsonFile = ExitOnErr(llvm::json::parse(Buf->getBuffer()));
+  llvm::json::Object *jsonContent = jsonFile.getAsObject();
+  llvm::json::Array *jsonArr = jsonContent->getArray(DEVICE_PLACEMENT_KEY);
+  if (!jsonArr || jsonArr->empty())
+    return;
+
+  // Collect operations to work on.
+  OpSetType workingOps(ops.begin(), ops.end());
+  // Go over operations in the JSON and find matched operation in the IR.
+  for (llvm::json::Value v : *jsonArr) {
+    llvm::json::Object *vobj = v.getAsObject();
+    StringRef device = vobj->getString(DEVICE_KEY).value();
+    StringRef nodeType = vobj->getString(NODE_TYPE_KEY).value();
+    StringRef nodeName = vobj->getString(ONNX_NODE_NAME_KEY).value();
+    LLVM_DEBUG(llvm::dbgs()
+               << "device: " << device.str() << ", nodeType: " << nodeType.str()
+               << ", nodeName: " << nodeName.str() << "\n");
+    OpSetType updatedOps;
+    for (Operation *op : workingOps) {
+      StringRef opNodeType = op->getName().getStringRef();
+      StringRef opNodeName =
+          op->getAttrOfType<mlir::StringAttr>("onnx_node_name").getValue();
+      // Match operation.
+      if (!std::regex_match(opNodeType.str(), std::regex(nodeType.str())))
+        continue;
+      if (!std::regex_match(opNodeName.str(), std::regex(nodeName.str())))
+        continue;
+      // Set device.
+      op->setAttr(
+          DEVICE_ATTRIBUTE, StringAttr::get(module.getContext(), device));
+      updatedOps.insert(op);
+    }
+    // To reduce complexity, once an operation is assigned a device, we remove
+    // it from the set workingOps.
+    workingOps = llvm::set_difference(workingOps, updatedOps);
+  }
+}
+
+void DevicePlacementPass::saveConfigToJSONFile() {
+  // Parsing the module to JSON object.
+  llvm::json::Array jsonArr;
+  for (Operation *op : ops) {
+    // Create a JSON object for this operation.
+    std::string deviceStr =
+        op->getAttrOfType<mlir::StringAttr>("device")
+            ? op->getAttrOfType<mlir::StringAttr>("device").getValue().str()
+            : "";
+    std::string nodeTypeStr = op->getName().getStringRef().str();
+    std::string nodeNameStr =
+        op->getAttrOfType<mlir::StringAttr>("onnx_node_name")
+            ? op->getAttrOfType<mlir::StringAttr>("onnx_node_name")
+                  .getValue()
+                  .str()
+            : "";
+    llvm::json::Value jsonObj = llvm::json::Object{
+        {DEVICE_KEY, deviceStr},
+        {NODE_TYPE_KEY, nodeTypeStr},
+        {ONNX_NODE_NAME_KEY, nodeNameStr},
+    };
+    jsonArr.emplace_back(jsonObj);
+  }
```

Comment:
Is it possible to save attributes to Json file only when the attribute is not ""? Should we save  the attribute ""?

Reply:
Yes, we should because it helps users know the `node_type` and `onnx_node_name` of the operation, and use these information to decide whether they want the operation to run on CPU by changing `device` from "" to "cpu". Otherwise, users need some way to explore these information, e.g. using --EmitONNXIR.